[WIP]: Refactoring and implementing unified tableau structure and generic perform_step! for SDIRK solver set

KrishGaur1354 opened the pull request:
This PR refactors the SDIRK solver implementation to use a unified tableau structure (`SDIRKTableau`) 
and a single generic `perform_step!` function, consolidating 95% of SDIRK methods and reducing code 
duplication. All methods now share consistent tableau coefficients while maintaining 
full backward compatibility and performance. Special handling preserved for TRBDF2's unique structure. 
KenCarp/Kvaerno methods retain separate implementations for additive splitting complexity.

Remaining Work:
KenCarp/Kvaerno Generic Integration - Migrate additive splitting methods to generic perform_step!
Duplication Removal of SDIRK related code
Re-run tests to check if working proper

Part of [SciML Small Grants Program](https://github.com/SciML/sciml.ai/blob/master/small_grants.md#refactor-ordinarydiffeqjl-solver-sets-to-reuse-perform_step-implementations-via-tableaus-100solver-set) for OrdinaryDiffEq.jl solver set refactoring

KrishGaur1354 commented:
@ChrisRackauckas could you provide any suggestions or any changes I should make?

oscardssmith commented on lib/OrdinaryDiffEqSDIRK/src/generic_sdirk_perform_step.jl:
```diff
@@ -0,0 +1,601 @@
+end
+
+@muladd function perform_step!(integrator, cache::SFSDIRK8Cache, repeat_step=false)
+    generic_sdirk_perform_step!(integrator, cache, repeat_step)
+end
+
+@muladd function perform_step!(integrator, cache::ESDIRK54I8L2SAConstantCache, repeat_step=false)
+    generic_sdirk_perform_step!(integrator, cache, repeat_step)
+end
+
+@muladd function perform_step!(integrator, cache::ESDIRK54I8L2SACache, repeat_step=false)
+    generic_sdirk_perform_step!(integrator, cache, repeat_step)
+end
+
+@muladd function perform_step!(integrator, cache::ESDIRK436L2SA2ConstantCache, repeat_step=false)
+    generic_sdirk_perform_step!(integrator, cache, repeat_step)
+end
+
+@muladd function perform_step!(integrator, cache::ESDIRK436L2SA2Cache, repeat_step=false)
+    generic_sdirk_perform_step!(integrator, cache, repeat_step)
+end
+
+@muladd function perform_step!(integrator, cache::ESDIRK437L2SAConstantCache, repeat_step=false)
+    generic_sdirk_perform_step!(integrator, cache, repeat_step)
+end
+
+@muladd function perform_step!(integrator, cache::ESDIRK437L2SACache, repeat_step=false)
+    generic_sdirk_perform_step!(integrator, cache, repeat_step)
+end
+
+@muladd function perform_step!(integrator, cache::ESDIRK547L2SA2ConstantCache, repeat_step=false)
+    generic_sdirk_perform_step!(integrator, cache, repeat_step)
+end
+
+@muladd function perform_step!(integrator, cache::ESDIRK547L2SA2Cache, repeat_step=false)
+    generic_sdirk_perform_step!(integrator, cache, repeat_step)
+end
+
+@muladd function perform_step!(integrator, cache::ESDIRK659L2SAConstantCache, repeat_step=false)
+    generic_sdirk_perform_step!(integrator, cache, repeat_step)
```
can all these caches be unified? This amount of repetition shouldn't be necessary.

KrishGaur1354 replied:
Actually, it was a bit easier for me to test it by repeating it the caches. Sure, I'll unify the repetitive perform_step! functions and remove the old implementations as suggested.

oscardssmith commented:
The primary thing this PR seems to be missing is a bunch of deletions. If this PR unifies the SDIRK tableau structure and perform steps, why isn't it deleting the old tableaus and perform step methods?

KrishGaur1354 commented:
> What's the unique structure? TRBDF2 and KenCarps should be able to join just fine. Any step predictor here is just a linear combination of previous values, which these match, so you just need an alpha array of step predictor coefficients.

Sure, I'll just ensure that the TRBDF2 and KenCarp methods would now be unified under consistent structure, which will utilize shared tableau and step predictor coefficients for it.

oscardssmith commented on lib/OrdinaryDiffEqSDIRK/src/unified_sdirk_tableaus.jl:
```diff
@@ -0,0 +1,721 @@
+    d = T(1 - sqrt(2) / 2) 
+    ω = T(sqrt(2) / 4)
+    
+    A = @SMatrix [0  0  0;
```
are static arrays actually faster here than just normal Vector/Matrix?

KrishGaur1354 replied:
Yes, I did found static arrays being significantly faster than normal Vector/Matrix for SDIRK tableau operations by almost 1.8 times faster.

oscardssmith replied:
That's interesting (and somewhat unfortunate). I would have expected the nonlinear solve cost to dominate.

ChrisRackauckas commented on lib/OrdinaryDiffEqSDIRK/src/kencarp_kvaerno_caches.jl:
```diff
@@ -7,7 +7,7 @@ function alg_cache(alg::Kvaerno3, u, rate_prototype, ::Type{uEltypeNoUnits},
         uprev, uprev2, f, t, dt, reltol, p, calck,
         ::Val{false}) where {uEltypeNoUnits, uBottomEltypeNoUnits, tTypeNoUnits}
-    tab = Kvaerno3Tableau(constvalue(uBottomEltypeNoUnits), constvalue(tTypeNoUnits))
+    tab = get_sdirk_tableau(:Kvaerno3, constvalue(uBottomEltypeNoUnits), constvalue(tTypeNoUnits))
```
What's the reason for Symbol rather than alg dispatch here?

ChrisRackauckas commented on lib/OrdinaryDiffEqSDIRK/src/unified_sdirk_tableaus.jl:
```diff
@@ -0,0 +1,684 @@
+    A_explicit::Union{SMatrix{S, S, T}, Nothing}
+    b_explicit::Union{SVector{S, T}, Nothing}
+    c_explicit::Union{SVector{S, T2}, Nothing}
+end
```
the tableaus are missing a step predictor coefficients part?

ChrisRackauckas commented on lib/OrdinaryDiffEqSDIRK/src/generic_sdirk_perform_step.jl:
```diff
@@ -0,0 +1,478 @@
+        z₁ = dt * integrator.fsalfirst
+    end
+    
+    if s == 4  # 4-stage methods like KenCarp3, Kvaerno3
```
why this instead of just stages?

ChrisRackauckas commented on lib/OrdinaryDiffEqSDIRK/src/generic_sdirk_perform_step.jl:
```diff
@@ -0,0 +1,478 @@
+    end
+    
+    if s == 4  # 4-stage methods like KenCarp3, Kvaerno3
+        z₂ = z₁
```
This is not generally a good prediction? It should be a linear combination from tableau coefficients?

KrishGaur1354 commented:
@oscardssmith @ChrisRackauckas I would like to apologies for the delay from my side. I have made the necessary changes for the refactoring. Do let me know if there's anything else to be added/removed or if any change is required in this PR.

ChrisRackauckas commented on lib/OrdinaryDiffEqSDIRK/src/sdirk_caches.jl:
```diff
@@ -654,21 +691,17 @@ end
 end
 
-@cache mutable struct Hairer4Cache{uType, rateType, uNoUnitsType, Tab, N} <:
+@cache mutable struct Hairer4Cache{uType, rateType, uNoUnitsType, Tab, N, StepLimiter} <:
```
The point of the refactor is that these types don't need to exist. If there's a single SDIRKCache that just has different values for the tableau, then every SDIRK algorithm gives the same cache type, and so they compile all the same. If you have separate cache types, they will all compile different codes. So this doesn't meet the criteria. I recommend looking at the Rosenbrock methods, where a good chunk of them are all the same cache https://github.com/SciML/OrdinaryDiffEq.jl/blob/master/lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_caches.jl#L748-L832, and all the same perform_step!, and it's just written based on the tableau size. This makes it so that compiling for one Rosenbrock compiles for all of them, effectively removing the per-algorithm compilation.

ChrisRackauckas commented:
Test failure looks real.

ChrisRackauckas commented on lib/OrdinaryDiffEqSDIRK/src/generic_sdirk_perform_step.jl:
```diff
@@ -0,0 +1,606 @@
+            if hasproperty(tab, :α_pred) && tab.α_pred !== nothing
+                z₃ = tab.α_pred[3,1] * z₁ + tab.α_pred[3,2] * z₂
+            else
+                θ = c[3] / c[2]
+                α31 = ((1 + (-4θ + 3θ^2)) + (6θ * (1 - θ) / c[2]) * γ)
+                α32 = ((-2θ + 3θ^2)) + (6θ * (1 - θ) / c[2]) * γ
+                z₃ = α31 * z₁ + α32 * z₂
```
why is this hardcoded here instead of being in the tableau?

ChrisRackauckas commented on lib/OrdinaryDiffEqSDIRK/src/generic_sdirk_perform_step.jl:
```diff
@@ -0,0 +1,606 @@
+    generic_sdirk_perform_step!(integrator, cache, repeat_step)
+end
+
+@muladd function generic_additive_sdirk_perform_step!(integrator, cache::SDIRKConstantCache, repeat_step=false)
```
why is this kept separate? And what is meant by "additive SDRIK"?

KrishGaur1354 commented on lib/OrdinaryDiffEqSDIRK/src/generic_sdirk_perform_step.jl:
```diff
@@ -0,0 +1,606 @@
+end
+
+# TRBDF2 special handling for constant cache
+@muladd function perform_step_trbdf2!(integrator, cache::SDIRKConstantCache, repeat_step=false)
```
Removed TRBDF2 special handling, so it now uses the standard ESDIRK implementation through the unified generic function

KrishGaur1354 commented on lib/OrdinaryDiffEqSDIRK/src/tableau_utils.jl:
```diff
@@ -0,0 +1,269 @@
+    end
+end
+
+# simple 1-stage SDIRK perform step (ImplicitEuler, ImplicitMidpoint)
```
Removed specialized 1-stage and 2-stage functions, so the generic implementation handles all stage counts uniformly

ChrisRackauckas commented on lib/OrdinaryDiffEqSDIRK/src/generic_sdirk_perform_step.jl:
```diff
@@ -0,0 +1,358 @@
+    integrator.u = u
+end
+
+# Dispatch for unified caches - all SDIRK algorithms use these same cache types
+@muladd function perform_step!(integrator, cache::SDIRKCache, repeat_step=false)
+    generic_sdirk_perform_step!(integrator, cache, repeat_step)
+end
+
+@muladd function perform_step!(integrator, cache::SDIRKConstantCache, repeat_step=false)
+    generic_sdirk_perform_step!(integrator, cache, repeat_step)
```
Can just move the code into here instead of having a different function call?

ChrisRackauckas commented on lib/OrdinaryDiffEqSDIRK/src/unified_sdirk_tableaus.jl:
```diff
@@ -0,0 +1,742 @@
+    is_A_stable::Bool
+    is_L_stable::Bool
+    predictor_type::Symbol
+    A_explicit::Union{SMatrix{S, S, T}, Nothing}
```
It's not type-stable to grab these: should just parameterize this?

ChrisRackauckas commented on lib/OrdinaryDiffEqSDIRK/src/unified_sdirk_tableaus.jl:
```diff
@@ -0,0 +1,742 @@
+                 predictor_type=:hermite)
+end
+
+function get_sdirk_tableau(alg::Symbol, T=Float64, T2=Float64)
```
```suggestion
function get_sdirk_tableau(alg::Symbol, ::Type{T}=Float64, ::Type{T2}=Float64) where {T,T2}
```

ChrisRackauckas commented:
Okay this is generally looking right. Left some comments. Probably just needs one more small round? Most are just little inference things. I think the only thing that's not trivial inference stuff is:

1. ImplicitEuler and Trapezoidal had the SPICE error estimators, how are those handled? I don't see that in there?
2. There are some methods that had built-in step predictors like TRBDF2 and KenCarps that I don't see changed to using the alphas? Curious if that was intentional.

ChrisRackauckas commented on lib/OrdinaryDiffEqSDIRK/src/unified_sdirk_tableaus.jl:
```diff
@@ -0,0 +1,757 @@
+                 predictor_type=:hermite)
+end
+
+function get_sdirk_tableau(alg::Symbol, ::Type{T}=Float64, ::Type{T2}=Float64) where {T, T2}
```
This is a bit wonky. It should instead just be like `SDIRKTableau(KenCarp4(), ...)` i.e. dispatch on the algorithm rather than different functions and this would all go away. But I can just refactor that at some later point. I wouldn't worry about this for now but it is a very weird way to do dispatch.

ChrisRackauckas commented on lib/OrdinaryDiffEqSDIRK/src/sdirk_caches.jl:
```diff
@@ -4,999 +4,105 @@ function get_fsalfirstlast(cache::SDIRKMutableCache, u)
-    SFSDIRK5Cache(u, uprev, fsalfirst, z₁, z₂, z₃, z₄, z₅, z₆, atmp, nlsolver, tab)
-end
+# Unified alg_cache functions for mutable cache
+function alg_cache(alg::Union{ImplicitEuler, ImplicitMidpoint, Trapezoid, TRBDF2, SDIRK2, SDIRK22, SSPSDIRK2,
```
Is there not an abstract type for all of the SDIRK methods to dispatch with? That seems much more sane than a big union

ChrisRackauckas commented on lib/OrdinaryDiffEqSDIRK/src/generic_sdirk_perform_step.jl:
```diff
@@ -0,0 +1,345 @@
+                @.. broadcast=false zi = dt * integrator.fsalfirst
+            else
+                fill!(zi, zero(eltype(u)))
+            end
```
Oh this is the first step part.

ChrisRackauckas commented:
No worries, take the time you need. Near the end now!

KrishGaur1354 commented:
> No worries, take the time you need. Near the end now!

So, @ChrisRackauckas I have addressed some of the things you recommended to look into:

- ImplicitEuler and Trapezoid SPICE error estimators which is now implemented with `isa` checks
- TRBDF2, KenCarp3, Kvaerno3 which now use α_pred predictors  
- Package compiles successfully

Place where I got stuck: Trapezoid and TRBDF2 fail on first step because SPICE estimators need `success_iter > 0`. With no history and no embedded method, they set `EEst = 1` causing immediate failure. So, how should these handle the first step before SPICE history exists?

ImplicitEuler and KenCarp4 work correctly. Once the first-step case is resolved, I can run full benchmarks.

ChrisRackauckas commented:
Rebase onto latest mater and tests should be passing.

KrishGaur1354 commented:
@ChrisRackauckas @oscardssmith Rebased onto latest master and fixed compatibility issues:
- Replaced @unpack with modern destructuring syntax (removed in Core)
- Fixed strip_cache for SDIRKCache (falls back to NamedTuple)
- All quick tests pass (Trapezoid, TRBDF2, strip tests)

ChrisRackauckas commented:
Rebase onto master, tests should be fixed there so this should just show up as passing and merge.
